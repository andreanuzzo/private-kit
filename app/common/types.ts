import { Posix } from '../helpers/dateTimeUtils';

export type LocalDataPoint = {
  latitude: number;
  longitude: number;
  time: Posix;
  hashes: string[];
};

export type Coordinates = {
  latitude: number;
  longitude: number;
};

export type ApiHealthcareAuthority = {
  name: string;
  bounds: {
    ne: Coordinates;
    sw: Coordinates;
  };
  org_id: string;
  cursor_url: string;
  public_api: string;
};

export type HealthcareAuthority = ApiHealthcareAuthority & {
  internal_id: string;
};
