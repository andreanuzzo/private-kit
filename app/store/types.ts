import { store } from './index';
import rootReducer from './reducers/rootReducer';
import { StateType } from 'typesafe-actions';
import { ThunkAction } from 'redux-thunk';
import { Action } from 'redux';
// Our internal store representation is the same as the API representation
export type { HealthcareAuthority } from '../common/types';

export type RootState = ReturnType<typeof rootReducer>;
export type Store = StateType<typeof store>;

// These will appear in redux debugger, so string names are helpful
export enum ApiStatus {
  INITIAL = 'INITIAL',
  STARTED = 'STARTED',
  SUCCESS = 'SUCCESS',
  FAILURE = 'FAILURE',
}

export type ApiRequest = {
  status: ApiStatus; // This allows components to show states based on api status
  errorMessage: string | null; // This is only for redux debugging. Store as a string for the safety in the store.
};

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export enum FeatureFlagOption {
  IMPORT_LOCATIONS_GOOGLE = 'IMPORT_LOCATIONS_GOOGLE',
  IMPORT_LOCATIONS_JSON_URL = 'IMPORT_LOCATIONS_JSON_URL',
  CUSTOM_URL = 'CUSTOM_URL',
  DOWNLOAD_LOCALLY = 'DOWNLOAD_LOCALLY',
  DEV_LANGUAGES = 'DEV_LANGUAGES',
  BYPASS_EXPORT_API = 'BYPASS_EXPORT_API',
  MOCK_EXPOSURE = 'MOCK_EXPOSURE',
}
