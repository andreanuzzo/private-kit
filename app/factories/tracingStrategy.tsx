import React from 'react';
import { View } from 'react-native';
import { Factory } from 'fishery';

import { TracingStrategy, StrategyAssets } from '../tracingStrategy';

import { Images } from '../../app/assets/images';

export default Factory.define<TracingStrategy>(() => ({
  name: 'test-tracing-strategy',
  exposureEventsStrategy: {
    exposureInfoSubscription: () => {
      return { remove: () => {} };
    },
    toExposureHistory: () => [],
    getCurrentExposures: () => {},
  },
  permissionsProvider: PermissionsProvider,
  homeScreenComponent: HomeScreen,
  affectedUserFlow: AffectedUserFlow,
  assets: testStrategyAssets,
}));

const PermissionsProvider = ({
  children,
}: {
  children: JSX.Element;
}): JSX.Element => {
  return <View testID={'permissions-provider'}>{children}</View>;
};

const HomeScreen = () => {
  return <View testID={'home-screen'} />;
};

const AffectedUserFlow = () => {
  return <View testID={'affected-user-flow'} />;
};

export const testStrategyAssets: StrategyAssets = {
  personalPrivacyBackground: Images.BlueGradientBackground,
  personalPrivacyIcon: '',
  notificationDetailsBackground: Images.BlueGradientBackground,
  notificationDetailsIcon: '',
  shareDiagnosisBackground: Images.BlueGradientBackground,
  shareDiagnosisIcon: '',
};
