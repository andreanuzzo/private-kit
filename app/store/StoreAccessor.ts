import { Store } from 'redux';

/*
 * This accessor is added for the SOLE purpose of avoiding cyclic imports from
 * the IntersectService to the Intersect helper and again to the store. The
 * accessor serves as a bridge for the consumers.
 * */
class StoreAccessor {
  store: Store | null = null;

  setStore = (store: Store | null) => {
    this.store = store;
  };

  getStore = () => {
    return this.store;
  };
}

const singleton = new StoreAccessor();

export default singleton;
