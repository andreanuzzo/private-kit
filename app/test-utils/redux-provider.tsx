// test-utils.js
import React, { ReactElement, ReactNode } from 'react';
import {
  render as rtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react-native';
import { Store, createStore } from 'redux';
import { Provider, DefaultRootState } from 'react-redux';
import rootReducer from '../store/reducers/rootReducer';

type ReduxRenderOptions = RenderOptions & {
  initialState: DefaultRootState;
  store?: Store;
};

type ReduxProviderProps = {
  children?: ReactNode;
};

function render(
  ui: ReactElement<unknown>,
  {
    initialState,
    store = createStore(rootReducer, initialState),
    ...rest
  }: ReduxRenderOptions,
): RenderResult {
  function ReduxProvider({ children }: ReduxProviderProps) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: ReduxProvider, ...rest });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
