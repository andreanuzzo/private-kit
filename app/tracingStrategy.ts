import { ImageSourcePropType } from 'react-native';

import { ExposureEventsStrategy } from './ExposureHistoryContext';

export interface TracingStrategy {
  name: string;
  exposureEventsStrategy: ExposureEventsStrategy;
  permissionsProvider: ({ children }: { children: JSX.Element }) => JSX.Element;
  homeScreenComponent: ({ testID }: { testID: string }) => JSX.Element;
  affectedUserFlow: () => JSX.Element;
  assets: StrategyAssets;
}

export interface StrategyAssets {
  personalPrivacyBackground: ImageSourcePropType;
  notificationDetailsBackground: ImageSourcePropType;
  shareDiagnosisBackground: ImageSourcePropType;
  personalPrivacyIcon: string;
  notificationDetailsIcon: string;
  shareDiagnosisIcon: string;
}
