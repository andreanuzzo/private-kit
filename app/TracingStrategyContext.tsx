import React, { createContext, useContext, FunctionComponent } from 'react';

import { TracingStrategy, StrategyAssets } from './tracingStrategy';

import { ExposureHistoryProvider } from './ExposureHistoryContext';

interface TracingStrategyContextState {
  name: string;
  homeScreenComponent: ({ testID }: { testID: string }) => JSX.Element;
  affectedUserFlow: () => JSX.Element;
  assets: StrategyAssets;
}

const TracingStrategyContext = createContext<
  TracingStrategyContextState | undefined
>(undefined);

interface TracingStrategyProps {
  strategy: TracingStrategy;
}

export const TracingStrategyProvider: FunctionComponent<TracingStrategyProps> = ({
  children,
  strategy,
}) => {
  const StrategyPermissionsProvider = strategy.permissionsProvider;

  return (
    <TracingStrategyContext.Provider
      value={{
        name: strategy.name,
        homeScreenComponent: strategy.homeScreenComponent,
        affectedUserFlow: strategy.affectedUserFlow,
        assets: strategy.assets,
      }}>
      <StrategyPermissionsProvider>
        <ExposureHistoryProvider
          exposureEventsStrategy={strategy.exposureEventsStrategy}>
          {children}
        </ExposureHistoryProvider>
      </StrategyPermissionsProvider>
    </TracingStrategyContext.Provider>
  );
};

export const useTracingStrategyContext = (): TracingStrategyContextState => {
  const context = useContext(TracingStrategyContext);
  if (context === undefined) {
    throw new Error('TracingStrategyContext must be used with a provider');
  }
  return context;
};

export const useStrategyContent = (): {
  StrategyAssets: StrategyAssets;
} => {
  const { assets } = useTracingStrategyContext();
  const StrategyAssets = assets;
  return { StrategyAssets };
};
