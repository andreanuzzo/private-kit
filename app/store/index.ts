import AsyncStorage from '@react-native-community/async-storage';
import { createImmutableStateInvariantMiddleware } from '@reduxjs/toolkit';
import { applyMiddleware, createStore } from 'redux';
import { createMigrate, persistReducer, persistStore } from 'redux-persist';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';

import StoreAccessor from './StoreAccessor';
import onChangedSelectedHealthAuthorities from './middleware/onChangedSelectedHealthAuthorities';

import migrations from './migrations';
import rootReducer from './reducers/rootReducer';

const enhancers = composeWithDevTools(
  applyMiddleware(
    thunk,
    createImmutableStateInvariantMiddleware(),
    onChangedSelectedHealthAuthorities,
  ),
);

export const STORE_VERSION = 2;
const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
  version: STORE_VERSION,
  // add keys of reducers to ignore:
  blacklist: [],
  migrate: createMigrate(migrations),
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = createStore(persistedReducer, {}, enhancers);
StoreAccessor.setStore(store);
export const persistor = persistStore(store);
