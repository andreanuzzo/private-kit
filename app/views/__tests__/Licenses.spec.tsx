import React from 'react';
import 'react-native';
import { render } from '../../test-utils/redux-provider';
import '@testing-library/jest-native/extend-expect';
import { useNavigation, useFocusEffect } from '@react-navigation/native';

import { TracingStrategyProvider } from '../../TracingStrategyContext';
import { TracingStrategy } from '../../tracingStrategy';
import gpsStrategy from '../../gps';

import { LicensesScreen } from '../Licenses';
import { FeatureFlagOption } from '../../store/types';

const renderTracingStrategyProvider = (strategy: TracingStrategy) => {
  return render(
    <TracingStrategyProvider strategy={strategy}>
      <LicensesScreen />
    </TracingStrategyProvider>,
    {
      initialState: {
        featureFlags: { flags: { [FeatureFlagOption.MOCK_EXPOSURE]: false } },
      },
    },
  );
};

jest.mock('@react-navigation/native');
(useNavigation as jest.Mock).mockReturnValue({ navigate: jest.fn() });
(useFocusEffect as jest.Mock).mockReturnValue({ navigate: jest.fn() });

describe('LicensesScreen', () => {
  it('renders correctly', () => {
    const { asJSON } = renderTracingStrategyProvider(gpsStrategy);
    expect(asJSON()).toMatchSnapshot();
  });

  describe('when the tracing strategy is gps', () => {
    it('displays PathCheck GPS', () => {
      const { getByTestId } = renderTracingStrategyProvider(gpsStrategy);

      expect(getByTestId('licenses-legal-header')).toHaveTextContent(
        'PathCheck GPS',
      );
    });
  });
});
