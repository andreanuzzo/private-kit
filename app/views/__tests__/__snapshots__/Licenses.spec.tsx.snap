// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LicensesScreen renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <SafeAreaView
    style={
      Object {
        "backgroundColor": "#1f2c9b",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#1f2c9b",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "paddingHorizontal": 16,
          "paddingTop": undefined,
          "paddingVertical": 8,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
        >
          <SvgXml
            color="#ffffff"
            style={
              Object {
                "paddingTop": 12,
              }
            }
            xml="
  <svg width=\\"16\\" height=\\"27\\" viewBox=\\"0 0 16 27\\"  xmlns=\\"http://www.w3.org/2000/svg\\">
  <path d=\\"M14.6216 0.226764L15.0632 0.650056C15.3787 0.952408 15.3787 1.3757 15.0632 1.67805L2.76011 13.4698L15.0632 25.3219C15.3787 25.6243 15.3787 26.0476 15.0632 26.3499L14.6216 26.7732C14.3061 27.0756 13.8644 27.0756 13.549 26.7732L0.236403 14.014C-0.0790606 13.7116 -0.0790606 13.2884 0.236403 12.986L13.549 0.226764C13.8644 -0.0755879 14.3061 -0.0755879 14.6216 0.226764Z\\" fill=\\"currentColor\\"/>
  </svg>
"
          />
        </TouchableOpacity>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 5,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#2e2e2e",
                "fontFamily": "IBMPlexSans",
                "fontSize": 17,
                "lineHeight": 24,
              },
              Object {
                "writingDirection": "ltr",
              },
              Object {
                "color": "#ffffff",
                "fontFamily": "IBMPlexSans-Bold",
                "fontSize": 19,
                "fontWeight": "500",
                "lineHeight": 32,
                "textTransform": "uppercase",
              },
            ]
          }
        >
          Legal
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#f8f8ff",
          "flex": 1,
        }
      }
    >
      <ScrollView
        alwaysBounceVertical={false}
        contentContainerStyle={
          Object {
            "backgroundColor": "#f8f8f8",
            "paddingHorizontal": 24,
            "paddingTop": 24,
          }
        }
      >
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#1f2c9b",
                  "fontFamily": "IBMPlexSans-Bold",
                  "fontSize": 28,
                  "fontWeight": "500",
                  "lineHeight": 32,
                },
                Object {
                  "writingDirection": "ltr",
                },
                undefined,
              ]
            }
            testID="licenses-legal-header"
          >
            PathCheck GPS
          </Text>
          <View
            style={
              Object {
                "paddingLeft": 20,
                "paddingTop": 12,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#4e4e4e",
                    "fontFamily": "IBMPlexSans",
                    "fontSize": 17,
                    "lineHeight": 24,
                  },
                  Object {
                    "writingDirection": "ltr",
                  },
                  undefined,
                ]
              }
            >
              PathCheck Foundation
58 Day Street
Box 441621
Somerville, MA 02144
            </Text>
            <View
              style={
                Object {
                  "height": 20,
                }
              }
            />
            <Text
              style={
                Array [
                  Object {
                    "color": "#4e4e4e",
                    "fontFamily": "IBMPlexSans",
                    "fontSize": 17,
                    "lineHeight": 24,
                  },
                  Object {
                    "writingDirection": "ltr",
                  },
                  Object {
                    "color": "#1f2c9b",
                    "textDecorationLine": "underline",
                  },
                ]
              }
            >
              info@pathcheck.org
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#4e4e4e",
                    "fontFamily": "IBMPlexSans",
                    "fontSize": 17,
                    "lineHeight": 24,
                  },
                  Object {
                    "writingDirection": "ltr",
                  },
                  Object {
                    "color": "#1f2c9b",
                    "textDecorationLine": "underline",
                  },
                ]
              }
            >
              pathcheck.org
            </Text>
          </View>
        </View>
      </ScrollView>
      <TouchableOpacity
        activeOpacity={0.2}
        style={
          Object {
            "backgroundColor": "#4051db",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingHorizontal": 15,
            "paddingVertical": 25,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#2e2e2e",
                "fontFamily": "IBMPlexSans",
                "fontSize": 17,
                "lineHeight": 24,
              },
              Object {
                "writingDirection": "ltr",
              },
              Object {
                "color": "#ffffff",
              },
            ]
          }
        >
          Privacy Policy
        </Text>
        <View
          style={
            Object {
              "alignSelf": "center",
              "paddingLeft": 20,
              "paddingRight": 20,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../app/assets/images/foreArrow.png",
              }
            }
          />
        </View>
      </TouchableOpacity>
    </View>
  </SafeAreaView>
</View>
`;
